src/hotspot/share/oops/method.hpp

```diff
@@ -847,12 +848,9 @@ class Method : public Metadata {
   void release_C_heap_structures();
 
   Method* get_new_method() const {
-    InstanceKlass* holder = method_holder();
-    Method* new_method = holder->method_with_idnum(orig_method_idnum());
-
-    assert(new_method != nullptr, "method_with_idnum() should not be null");
+    Method* new_method = method_holder()->method_with_idnum(orig_method_idnum());
     assert(this != new_method, "sanity check");
-    return new_method;
+    return (new_method == nullptr || is_deleted()) ? Universe::throw_no_such_method_error() : new_method;
```

Comment:
I am still confused by the different possibilities here. Under what conditions will we get nullptr? Is it the case that `get_new_method` should only be called when `is_old()` is true? Can `is_old` and `is_deleted` be true at the same time?

Reply:
To answer some of my own questions:
- yes `get_new_method` should only be called if `is_old` is true. (Should we assert that?)
- yes a method can be old and deleted at the same time.

I remain unclear how nullptr can appear here.

Reply:
This redefine code is really complicated.  Obsolete methods, which includes deleted methods get an incremented idnum in check_methods_and_mark_as_obsolete.

        // obsolete methods need a unique idnum so they become new entries in
        // the jmethodID cache in InstanceKlass
        assert(old_method->method_idnum() == new_method->method_idnum(), "must match");
        u2 num = InstanceKlass::cast(_the_class)->next_method_idnum();
        if (num != ConstMethod::UNSET_IDNUM) {
          old_method->set_method_idnum(num);
        }

When get_new_method() is called it compares InstanceKlass::_methods[idnum], then compares the Method at that index to the idnum that the method has stored.  For deleted methods, this will return nullptr because the idnum comparison in method_with_idnum will not match.  I think it is sufficient to check for null for deleted methods but also the explicit comparison is much easier to understand that it's the right answer.

Yes, there could be an assert in get_new_method(), the method passed in is_old().  All redefined methods are marked as is_old().  I believe all callers test this before making this call.

Reply:
So is it the case that `nullptr` implies deleted, and deleted implies `nullptr`? If so checking for both is redundant and confusing because it makes it look like there are two distinct cases.

Reply:
So that implies that you trust my reading of this code.  It is complicated enough that testing both seems like a safe thing to do and somewhat clarifying, or else adding an assert like:

    assert(new_method != nullptr || old_method->is_deleted(), "this is the only way this happens");
    return new_method == nullptr ? nsme : new_method;

Reply:
The assert works for me.

Reply:
Can we assert the stronger statement: `(new_method == nullptr) == (old_method->is_deleted())` ?

Reply:
Yeah I like that better too.